File: sdks/typescript/src/utils/cache.ts
/**
 * LRU Cache implementation for memory operations
 * Production-ready with TTL support and storage backends
 */

import type { CacheConfig, CacheStorage } from '../types';

interface CacheEntry<T> {
  value: T;
  expiresAt: number;
  accessCount: number;
  lastAccessed: number;
}

/**
 * LRU Cache with TTL support
 */
export class LRUCache<T = unknown> {
  private cache: Map<string, CacheEntry<T>>;
  private maxSize: number;
  private ttl: number;
  private storage: CacheStorage;
  private storageKey = 'xache_cache';

  constructor(config: Required<CacheConfig>) {
    this.maxSize = config.maxSize;
    this.ttl = config.ttl;
    this.storage = config.storage;
    this.cache = new Map();

    if (this.storage === 'localStorage' && this.isLocalStorageAvailable()) {
      this.loadFromLocalStorage();
    }
  }

  /**
   * Check if localStorage is available (browser environment)
   */
  private isLocalStorageAvailable(): boolean {
    try {
      const testKey = '__xache_test__';
      const storage = (globalThis as any).localStorage;
      if (!storage) return false;
      storage.setItem(testKey, 'test');
      storage.removeItem(testKey);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Get value from cache
   */
  get(key: string): T | undefined {
    const entry = this.cache.get(key);

    if (!entry) {
      return undefined;
    }

    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      this.persistToLocalStorage();
      return undefined;
    }

    entry.accessCount++;
    entry.lastAccessed = Date.now();

    this.cache.delete(key);
    this.cache.set(key, entry);

    return entry.value;
  }

  /**
   * Set value in cache
   */
  set(key: string, value: T): void {
    if (this.cache.size >= this.maxSize && !this.cache.has(key)) {
      this.evictLRU();
    }

    const entry: CacheEntry<T> = {
      value,
      expiresAt: Date.now() + this.ttl,
      accessCount: 1,
      lastAccessed: Date.now(),
    };

    this.cache.set(key, entry);
    this.persistToLocalStorage();
  }

  /**
   * Check if key exists and is not expired
   */
  has(key: string): boolean {
    return this.get(key) !== undefined;
  }

  /**
   * Delete specific key
   */
  delete(key: string): boolean {
    const result = this.cache.delete(key);
    this.persistToLocalStorage();
    return result;
  }

  /**
   * Clear entire cache
   */
  clear(): void {
    this.cache.clear();
    this.persistToLocalStorage();
  }

  /**
   * Get cache statistics
   */
  getStats(): {
    size: number;
    maxSize: number;
    hitRate?: number;
  } {
    return {
      size: this.cache.size,
      maxSize: this.maxSize,
    };
  }

  /**
   * Evict least recently used entry
   */
  private evictLRU(): void {
    let oldestKey: string | undefined;
    let oldestTime = Infinity;

    for (const [key, entry] of this.cache.entries()) {
      if (entry.lastAccessed < oldestTime) {
        oldestTime = entry.lastAccessed;
        oldestKey = key;
      }
    }

    if (oldestKey) {
      this.cache.delete(oldestKey);
    }
  }

  /**
   * Load cache from localStorage
   */
  private loadFromLocalStorage(): void {
    if (this.storage !== 'localStorage' || !this.isLocalStorageAvailable()) {
      return;
    }

    try {
      const storage = (globalThis as any).localStorage;
      const stored = storage.getItem(this.storageKey);
      if (!stored) return;

      const data = JSON.parse(stored) as Array<[string, CacheEntry<T>]>;
      const now = Date.now();

      for (const [key, entry] of data) {
        if (entry.expiresAt > now) {
          this.cache.set(key, entry);
        }
      }
    } catch (error) {
      console.warn('Failed to load cache from localStorage:', error);
    }
  }

  /**
   * Persist cache to localStorage
   */
  private persistToLocalStorage(): void {
    if (this.storage !== 'localStorage' || !this.isLocalStorageAvailable()) {
      return;
    }

    try {
      const storage = (globalThis as any).localStorage;
      const data = Array.from(this.cache.entries());
      storage.setItem(this.storageKey, JSON.stringify(data));
    } catch (error) {
      console.warn('Failed to persist cache to localStorage:', error);
    }
  }

  /**
   * Clean up expired entries (run periodically)
   */
  cleanup(): void {
    const now = Date.now();
    const keysToDelete: string[] = [];

    for (const [key, entry] of this.cache.entries()) {
      if (entry.expiresAt <= now) {
        keysToDelete.push(key);
      }
    }

    for (const key of keysToDelete) {
      this.cache.delete(key);
    }

    if (keysToDelete.length > 0) {
      this.persistToLocalStorage();
    }
  }
}

File: sdks/python/xache/utils/cache.py
"""
LRU Cache implementation for memory operations
Production-ready with TTL support and pickle persistence
"""

import time
import pickle
import os
from typing import Dict, Generic, TypeVar, Optional, Any
from dataclasses import dataclass

T = TypeVar('T')


@dataclass
class CacheEntry(Generic[T]):
    """Cache entry with metadata"""
    value: T
    expires_at: float
    access_count: int
    last_accessed: float


class CacheConfig:
    """Configuration for cache behavior"""

    def __init__(
        self,
        enabled: bool = True,
        max_size: int = 100,
        ttl: int = 300000,  # 5 minutes in milliseconds
        storage: str = 'memory',  # 'memory' or 'pickle'
        pickle_path: Optional[str] = None
    ):
        self.enabled = enabled
        self.max_size = max_size
        self.ttl = ttl
        self.storage = storage
        self.pickle_path = pickle_path or os.path.expanduser('~/.xache_cache.pkl')


class LRUCache(Generic[T]):
    """LRU Cache with TTL support and pickle persistence"""

    def __init__(self, config: CacheConfig):
        self.max_size = config.max_size
        self.ttl = config.ttl / 1000  # Convert to seconds
        self.storage = config.storage
        self.pickle_path = config.pickle_path
        self.cache: Dict[str, CacheEntry[T]] = {}

        if self.storage == 'pickle':
            self._load_from_pickle()

    def get(self, key: str) -> Optional[T]:
        """
        Get value from cache

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found/expired
        """
        entry = self.cache.get(key)

        if entry is None:
            return None

        # Check expiration
        if time.time() > entry.expires_at:
            del self.cache[key]
            self._persist_to_pickle()
            return None

        # Update LRU metadata
        entry.access_count += 1
        entry.last_accessed = time.time()

        # Move to end (LRU)
        del self.cache[key]
        self.cache[key] = entry

        return entry.value

    def set(self, key: str, value: T) -> None:
        """
        Set value in cache

        Args:
            key: Cache key
            value: Value to cache
        """
        # Evict if at capacity
        if len(self.cache) >= self.max_size and key not in self.cache:
            self._evict_lru()

        entry = CacheEntry(
            value=value,
            expires_at=time.time() + self.ttl,
            access_count=1,
            last_accessed=time.time()
        )

        self.cache[key] = entry
        self._persist_to_pickle()

    def has(self, key: str) -> bool:
        """
        Check if key exists and is not expired

        Args:
            key: Cache key

        Returns:
            True if key exists and is valid
        """
        return self.get(key) is not None

    def delete(self, key: str) -> bool:
        """
        Delete specific key

        Args:
            key: Cache key

        Returns:
            True if key was deleted
        """
        if key in self.cache:
            del self.cache[key]
            self._persist_to_pickle()
            return True
        return False

    def clear(self) -> None:
        """Clear entire cache"""
        self.cache.clear()
        self._persist_to_pickle()

    def get_stats(self) -> Dict[str, Any]:
        """
        Get cache statistics

        Returns:
            Dictionary with cache stats
        """
        return {
            'size': len(self.cache),
            'max_size': self.max_size
        }

    def cleanup(self) -> None:
        """Clean up expired entries"""
        now = time.time()
        keys_to_delete = [
            key for key, entry in self.cache.items()
            if entry.expires_at <= now
        ]

        for key in keys_to_delete:
            del self.cache[key]

        if keys_to_delete:
            self._persist_to_pickle()

    def _evict_lru(self) -> None:
        """Evict least recently used entry"""
        if not self.cache:
            return

        oldest_key = None
        oldest_time = float('inf')

        for key, entry in self.cache.items():
            if entry.last_accessed < oldest_time:
                oldest_time = entry.last_accessed
                oldest_key = key

        if oldest_key:
            del self.cache[oldest_key]

    def _load_from_pickle(self) -> None:
        """Load cache from pickle file"""
        if self.storage != 'pickle':
            return

        try:
            if os.path.exists(self.pickle_path):
                with open(self.pickle_path, 'rb') as f:
                    data = pickle.load(f)

                # Filter expired entries
                now = time.time()
                self.cache = {
                    key: entry for key, entry in data.items()
                    if entry.expires_at > now
                }
        except Exception as e:
            print(f"Warning: Failed to load cache from pickle: {e}")

    def _persist_to_pickle(self) -> None:
        """Persist cache to pickle file"""
        if self.storage != 'pickle':
            return

        try:
            # Ensure directory exists
            os.makedirs(os.path.dirname(self.pickle_path), exist_ok=True)

            with open(self.pickle_path, 'wb') as f:
                pickle.dump(self.cache, f)
        except Exception as e:
            print(f"Warning: Failed to persist cache to pickle: {e}")
